### PR TITLE
TYP: Add type stubs for stringdtype in np.char and np.strings

### PR DESCRIPTION
**Pull request by ngoldbaum**
This is definitely an improvement but there are still some issues.

@jorenham [pointed out](https://github.com/numpy/numpy/issues/25693#issuecomment-2380878615) (unfortunately after I pushed everything) that we can substantially simplify some of these overloads using a `TypeVar`.

I also am not able to make expression like `np.split(AR_S, ' ')` type check properly, since from the type-checker's perspective `' '` is convertible to a unicode array, but in reality in the way we set up the promoters for StringDType, this does return a StringDType array. I had to cheat for functions like that where normally people pass scalars but it's typed to be able to accept an array by passing in an `AR_S` in the reveal tests where the unicode tests pass in a scalar.

There were also a couple of functions that were missing tests and I've added them.

### CODE REVIEW

**Comment by jorenham** [comment]
Since `StringDType` has char code `"T"`, you could use `T_co` instead of `S_co` for it. 
That way, the one for bytes doesn't have to be renamed to `B_co`, additionally avoiding it being confused with `uint8` (which has char code `B`)

File: numpy/_typing/_array_like.py [hunk 1]
```diff
@@ -21,6 +21,7 @@
     str_,
     bytes_,
 )
+from numpy._core.multiarray import StringDType
```
Comment (jorenham) [inline numpy/_typing/_array_like.py:24]: I'm sure about the runtime, but at least in the stubs, `StringDType` is defined in `numpy.dtypes`. So if it's the same at runtime, and `numpy._core.multiarray` re-exports it, then it could be cleaner to import directly it from `np.dtypes`

File: numpy/typing/tests/data/reveal/strings.pyi [hunk 2]
```diff
@@ -1,132 +1,201 @@
...
+assert_type(np.strings.isupper(AR_B), npt.NDArray[np.bool])
 assert_type(np.strings.isupper(AR_S), npt.NDArray[np.bool])
 
 assert_type(np.strings.str_len(AR_U), npt.NDArray[np.int_])
+assert_type(np.strings.str_len(AR_B), npt.NDArray[np.int_])
 assert_type(np.strings.str_len(AR_S), npt.NDArray[np.int_])
+
+assert_type(np.strings.translate(AR_U, AR_U), npt.NDArray[np.str_])
+assert_type(np.strings.translate(AR_B, AR_B), npt.NDArray[np.bytes_])
+assert_type(np.strings.translate(AR_S, AR_S), np.ndarray[np_t._Shape, np.dtypes.StringDType])
```
Comment (jorenham) [inline numpy/typing/tests/data/reveal/strings.pyi:201]: I'm guessing that this red thingy is here because there's a trailing newline missing or something 🤔

**Comment by jorenham** [comment]
> I also am not able to make expression like `np.split(AR_S, ' ')` type check properly

Maybe I'm missing something, but the second argument of `numpy.split` should be an int or an array-like of ints according to the docs 🤔

File: numpy/_typing/_array_like.py (hunk 1 above)
Comment (ngoldbaum) [reply numpy/_typing/_array_like.py:24]: Agreed, for some reason the `np.dtypes` import breaks under reloading or lazy-loading. I'm not sure why it happens but it's a pre-existing issue that until now hasn't been problematic. I'd need to poke around in a C debugger to understand why it's happening, this was just a quick fix for the tests.

**Comment by jorenham** [comment]
> I also am not able to make expression like np.split(AR_S, ' ') type check properly, since from the type-checker's perspective ' ' is convertible to a unicode array, but in reality in the way we set up the promoters for StringDType, this does return a StringDType array. I had to cheat for functions like that where normally people pass scalars but it's typed to be able to accept an array by passing in an AR_S in the reveal tests where the unicode tests pass in a scalar.

Ah after applying my mental autocorrect function, I realized that you're talking about  `numpy.string.strip` 👍🏻 . I can kinda relate to this, as I once made the mistake to ask for a "stripper" instead of a "splitter" in a hardware store 😅.

Anyway, I don't think that the issue here has to do with the fact that `str_` array-likes also accept raw `str`s, because there's no overlap between `U_co` and `S_co` (the `T` / `StringDType` one).

```python
@overload
def strip(a: S_co, chars: None | S_co = ...) -> np.ndarray[_Shape, np.dtypes.StringDType]: ...
```

The `chars` parameter is of `None | S_co`, which resolves to `_SupportsArray[StringDType] | None`. 

When call `numpy.string.split(AR_S, " ")`, then `a` is inferred as an `ndarray` of `StringDType`, which matches the `a: S_co` parameter of this *only* overload. 
So the typechecker won't look at any other overloads at this point (so your pattern matching mindset is indeed the right way to look at it).
Now the 2nd argument " " is inferred as `chars: str`. But `str` can't be assigned to `None`, and also not to `_SupportsArray[StringDType]` (because `str` doesn't implement `__array__`).
At this point, there are no overloads left, and we're left stranded in the "land of undefined behavior".

So to make a short story long: the `chars` in the `StringDType` overload should probably also accept a `str`.

File: numpy/_typing/_array_like.py [hunk 3]
```diff
@@ -148,6 +149,7 @@ def __array_function__(
     dtype[bytes_],
     bytes,
 ]
+_ArrayLikeString_co: TypeAlias = _SupportsArray[StringDType]
```
Comment (jorenham) [inline numpy/_typing/_array_like.py:152]: I believe the `_co` suffix refers to it being "coerceable to [the relevant dtype]". 
So in this case, `_co` suffix might be misleading.

Removing it would probably be the easiest option.

Otherwise, (nested sequences of) `builtins.str` and `numpy.str_` array-likes should also be included, because

```pycon
>>> np.array(np.array(["True", "False", "FileNotFound"], np.str_), dtype=np.dtypes.StringDType())
array(['True', 'False', 'FileNotFound'], dtype=StringDType())
```

You could also consider doing both,  i.e. remove the `_co` from the current alias, and add a new `_co` one, that also includes these `str` and `str_` array-likes.

This was, you could type the first param of (the last overload of)  `numpy.strings.index` as a `_ArrayLikeString`, and the second param as `_ArrayLikeString_co`.

File: numpy/_core/strings.pyi [hunk 4]
```diff
@@ -4,67 +4,88 @@ import numpy as np
...
+def isdigit(x: U_co | B_co | S_co) -> NDArray[np.bool]: ...
+def isspace(x: U_co | B_co | S_co) -> NDArray[np.bool]: ...
+def isdecimal(x: U_co | S_co) -> NDArray[np.bool]: ...
+def isnumeric(x: U_co | S_co) -> NDArray[np.bool]: ...
+def islower(a: U_co | B_co | S_co) -> NDArray[np.bool]: ...
+def istitle(a: U_co | B_co | S_co) -> NDArray[np.bool]: ...
+def isupper(a: U_co | B_co | S_co) -> NDArray[np.bool]: ...
 
-def str_len(x: U_co | S_co) -> NDArray[np.int_]: ...
+def str_len(x: U_co | B_co | S_co) -> NDArray[np.int_]: ...
```
Comment (jorenham) [inline numpy/_core/strings.pyi:88]: You could consider using `TypeAlias` to reduce some of the duplication here.

File: numpy/_core/defchararray.pyi [hunk 5]
```diff
@@ -463,100 +464,132 @@ __all__: list[str]
...
 
-# String operations
 @overload
-def add(x1: U_co, x2: U_co) -> NDArray[str_]: ...
+def add(x1: U_co, x2: U_co) -> NDArray[np.str_]: ...
+@overload
+def add(x1: B_co, x2: B_co) -> NDArray[np.bytes_]: ...
 @overload
-def add(x1: S_co, x2: S_co) -> NDArray[bytes_]: ...
+def add(x1: S_co, x2: S_co) -> np.ndarray[_Shape, np.dtypes.StringDType]: ...
```
Comment (jorenham) [inline numpy/_core/defchararray.pyi:511]: The "`TypeVar` with constraints" trick can't be used here unfortunately.

But you could consider using a `TypeAlias` for `np.ndarray[_Shape, np.dtypes.StringDType]` to make this a bit more DRY

File: numpy/_core/defchararray.pyi [hunk 6]
```diff
@@ -566,12 +599,19 @@ def replace(
     count: None | i_co = ...,
 ) -> NDArray[str_]: ...
 @overload
+def replace(
+    a: B_co,
+    old: B_co,
+    new: B_co,
+    count: None | i_co = ...,
+) -> NDArray[bytes_]: ...
```
Comment (jorenham) [inline numpy/_core/defchararray.pyi:607]: It's probably best for consistency to leave it like it is, but in case you didn't know already: 
The official typing docs recommend using a `130` character limit in `.pyi`, so I'm guessing that this could fit on one line.

https://typing.readthedocs.io/en/latest/guides/writing_stubs.html#maximum-line-length

File: numpy/_core/defchararray.pyi [hunk 7]
```diff
@@ -598,16 +646,24 @@ def rsplit(
...
+    sep: None | B_co = ...,
+    maxsplit: None | i_co = ...,
+) -> NDArray[object_]: ...
+@overload
 def rsplit(
     a: S_co,
     sep: None | S_co = ...,
     maxsplit: None | i_co = ...,
-) -> NDArray[object_]: ...
+) -> NDArray[np.object_]: ...
```
Comment (jorenham) [inline numpy/_core/defchararray.pyi:659]: I don't see why this `np.` is needed here

File: numpy/_core/defchararray.pyi [hunk 8]
```diff
@@ -616,31 +672,41 @@ def split(
...
+    sep: None | B_co = ...,
+    maxsplit: None | i_co = ...,
+) -> NDArray[object_]: ...
+@overload
 def split(
     a: S_co,
     sep: None | S_co = ...,
     maxsplit: None | i_co = ...,
-) -> NDArray[object_]: ...
+) -> NDArray[np.object_]: ...
```
Comment (jorenham) [inline numpy/_core/defchararray.pyi:685]: See my comment above

**Review by jorenham — CHANGES_REQUESTED** [review]

File: numpy/typing/tests/data/reveal/strings.pyi [hunk 9]
```diff
@@ -1,132 +1,201 @@
...
 
 assert_type(np.strings.index(AR_U, "a", start=[1, 2, 3]), npt.NDArray[np.int_])
-assert_type(np.strings.index(AR_S, [b"a", b"b", b"c"], end=9), npt.NDArray[np.int_])
+assert_type(np.strings.index(AR_B, [b"a", b"b", b"c"], end=9), npt.NDArray[np.int_])
+assert_type(np.strings.index(AR_S, AR_S, start=[1, 2, 3]), npt.NDArray[np.int_])
+
 assert_type(np.strings.rindex(AR_U, "a", start=[1, 2, 3]), npt.NDArray[np.int_])
-assert_type(np.strings.rindex(AR_S, [b"a", b"b", b"c"], end=9), npt.NDArray[np.int_])
+assert_type(np.strings.rindex(AR_B, [b"a", b"b", b"c"], end=9), npt.NDArray[np.int_])
+assert_type(np.strings.rindex(AR_S, AR_S, start=[1, 2, 3]), npt.NDArray[np.int_])
```
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply numpy/typing/tests/data/reveal/strings.pyi:159]: This is what I was referring to when I opened the PR about string constants seeming to be coerceable to unicode strings. Hopefully once I implement your other comment i'll be able to make it match the other tests.

File: numpy/typing/tests/data/reveal/char.pyi [hunk 10]
```diff
@@ -2,146 +2,209 @@ from typing import Any
...
 
 assert_type(np.char.split(AR_U, "_"), npt.NDArray[np.object_])
-assert_type(np.char.split(AR_S, maxsplit=[1, 2, 3]), npt.NDArray[np.object_])
+assert_type(np.char.split(AR_B, maxsplit=[1, 2, 3]), npt.NDArray[np.object_])
+assert_type(np.char.split(AR_S, AR_S), npt.NDArray[np.object_])
+
 assert_type(np.char.rsplit(AR_U, "_"), npt.NDArray[np.object_])
-assert_type(np.char.rsplit(AR_S, maxsplit=[1, 2, 3]), npt.NDArray[np.object_])
+assert_type(np.char.rsplit(AR_B, maxsplit=[1, 2, 3]), npt.NDArray[np.object_])
+assert_type(np.char.rsplit(AR_S, AR_S), npt.NDArray[np.object_])
```
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply numpy/typing/tests/data/reveal/char.pyi:107]: This goes back to a decision @lysnikolaou made before 2.0 came out. For functions where the output can't be an array, we're not sure it really makes sense to define them as ufuncs or support them in the future. Rather than making a decision we just decided not to add functions like this to `strings` until we can come up with a better solution. Maybe a padded split with a known output shape and that pads with empty strings?

File: numpy/typing/tests/data/reveal/char.pyi [hunk 11]
```diff
@@ -2,146 +2,209 @@ from typing import Any
...
+assert_type(np.char.str_len(AR_B), npt.NDArray[np.int_])
 assert_type(np.char.str_len(AR_S), npt.NDArray[np.int_])
 
+assert_type(np.char.translate(AR_U, AR_U), npt.NDArray[np.str_])
+assert_type(np.char.translate(AR_B, AR_B), npt.NDArray[np.bytes_])
+assert_type(np.char.translate(AR_S, AR_S), np.ndarray[np_t._Shape, np.dtypes.StringDType])
+
 assert_type(np.char.array(AR_U), np.char.chararray[tuple[int, ...], np.dtype[np.str_]])
-assert_type(np.char.array(AR_S, order="K"), np.char.chararray[tuple[int, ...], np.dtype[np.bytes_]])
+assert_type(np.char.array(AR_B, order="K"), np.char.chararray[tuple[int, ...], np.dtype[np.bytes_]])
```
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply numpy/typing/tests/data/reveal/char.pyi:200]: I'm surprised it does anything at all! The plan is to eventually deprecate and remove `chararray`, and we don't want to ever add support for StringDType to it. I guess in principle we could make these explicit errors, but I don't think it's a high priority to mess with `chararray`.

File: numpy/_typing/_array_like.py (hunk 3 above)
Comment (ngoldbaum) [reply numpy/_typing/_array_like.py:152]: The behavior you saw with `index` is a bug I think (probably a missing promoter). 

That said, if I want to use `_ArrayLikeString_co` and have it mean that *all* strings are coercible to StringDType, suddenly that becomes context dependent. One of the arguments has to be explicitly a StringDType array for the promoters to fire, otherwise NumPy would use the unicode DType loops. So, non-StringDType arguments *are* coercible, but not in the general case. Is there a way to express that in the Python type system?

File: numpy/typing/tests/data/reveal/char.pyi (hunk 11 above)
[the remark this answers is not in the document]
Comment (jorenham) [reply numpy/typing/tests/data/reveal/char.pyi:200]: Ah I see, that makes sense. Then once it is officially deprecated, `@deprecated` could also be added to these methods in the stubs

File: numpy/typing/tests/data/reveal/char.pyi (hunk 10 above)
[the remark this answers is not in the document]
Comment (jorenham) [reply numpy/typing/tests/data/reveal/char.pyi:107]: I haven't encountered any use-cases for it myself, so this is all rather new to me. 
When I used it for the first time yesterday, I just assumed that it'd be like a "vectorized `builtins.str.split`", which seemed to work as expected, and I didn't have to look up the docs for it.
So I guess my point is that it might indeed be a good idea to stay close to the `str.split` API 🤷🏻.

And I can see why returning an `object_` array is awkward, but necessary to allow for variable amount of splits. 

Padding seems like a good solution, perhaps with an optional `pad: str = ""` kwarg, so that the new output axis has the size of the max amount of splits. So e.g. `split(["a_b", "a_b_c"], "_")` would then return an array of shape `(2, 3)`, where `out[0, :] == ["a", "b", ""]` (and for `rsplit` you'd then get `["", "a", "b"]` I guess). I suppose that this requires `maxsplit: int`, maybe with `-1` as default to indicate "the longest of all splits".
But you won't be able to use a ufunc anymore in this case; so perhaps a gufunc then?

**Comment by ngoldbaum** [comment]
Argh, so it looks like the mypy tests are failing here and they didn't fail locally for me because this PR branch is based on `main` before #27419.

After #27419, this is no longer an error and I have no idea why:

```
FAILED numpy/typing/tests/test_typing.py::test_fail[strings] - AssertionError: Error mismatch at line 7
E           AssertionError: Error mismatch at line 7
E
E           Expression: np.char.equal(AR_U, AR_S)
E           Expected error: incompatible type
E           Observed error: ''
```

(Note that this is after fixing a logic error in `test_fail` which flipped the expected and observed error printouts)

File: numpy/typing/tests/data/reveal/char.pyi [hunk 12]
```diff
@@ -2,146 +2,209 @@ from typing import Any
...
 
 assert_type(np.char.index(AR_U, "a", start=[1, 2, 3]), npt.NDArray[np.int_])
-assert_type(np.char.index(AR_S, [b"a", b"b", b"c"], end=9), npt.NDArray[np.int_])
+assert_type(np.char.index(AR_B, [b"a", b"b", b"c"], end=9), npt.NDArray[np.int_])
+assert_type(np.char.index(AR_S, AR_S, start=[1, 2, 3]), npt.NDArray[np.int_])
+
 assert_type(np.char.rindex(AR_U, "a", start=[1, 2, 3]), npt.NDArray[np.int_])
-assert_type(np.char.rindex(AR_S, [b"a", b"b", b"c"], end=9), npt.NDArray[np.int_])
+assert_type(np.char.rindex(AR_B, [b"a", b"b", b"c"], end=9), npt.NDArray[np.int_])
+assert_type(np.char.rindex(AR_S, AR_S, start=[1, 2, 3]), npt.NDArray[np.int_])
```
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply numpy/typing/tests/data/reveal/char.pyi:155]: Opened a new issue to track this.

File: numpy/typing/tests/data/reveal/char.pyi [hunk 13]
```diff
@@ -2,146 +2,209 @@ from typing import Any
...
+assert_type(np.char.isupper(AR_B), npt.NDArray[np.bool])
 assert_type(np.char.isupper(AR_S), npt.NDArray[np.bool])
 
 assert_type(np.char.str_len(AR_U), npt.NDArray[np.int_])
+assert_type(np.char.str_len(AR_B), npt.NDArray[np.int_])
 assert_type(np.char.str_len(AR_S), npt.NDArray[np.int_])
 
+assert_type(np.char.translate(AR_U, AR_U), npt.NDArray[np.str_])
+assert_type(np.char.translate(AR_B, AR_B), npt.NDArray[np.bytes_])
+assert_type(np.char.translate(AR_S, AR_S), np.ndarray[np_t._Shape, np.dtypes.StringDType])
```
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply numpy/typing/tests/data/reveal/char.pyi]: I think the existing overloads are just wrong, the other two arguments should be `str` and are documented as such.

File: numpy/_core/defchararray.pyi [hunk 14]
```diff
@@ -464,66 +468,86 @@ __all__: list[str]
...
 def capitalize(a: S_co) -> NDArray[bytes_]: ...
+@overload
+def capitalize(a: T_co) -> AR_T: ...
 
 @overload
 def center(a: U_co, width: i_co, fillchar: U_co = ...) -> NDArray[str_]: ...
 @overload
 def center(a: S_co, width: i_co, fillchar: S_co = ...) -> NDArray[bytes_]: ...
+@overload
+def center(a: T_co, width: i_co, fillchar: T_co = ...) -> AR_T: ...
```
Comment (jorenham) [inline numpy/_core/defchararray.pyi:542]: Now that `T_co` includes more stuff, it overlaps with `U_co`. 
But these first and thirds overloads are incompatible return types.
Pyright/pylance reports this as e.g.

```
Overload 1 for "add" overlaps overload 3 and returns an incompatible type
```

This is a very common issue within the numpy stubs actually, and I've made it a couple of times myself. 
It's not really obvious to see why this is an issue, especially in this particular case.
So here's an example of what exactly can go wrong:

```python
vec_str: list[str]
vec_str_or_T: list[str] | ndarray[tuple[int], StringDType]

reveal_type(capitalize(vec_str))
reveal_type(capitalize(vec_str_or_T))
```

and the (pyright) output is

```
Type of "capitalize(vec_str)" is "ndarray[tuple[int, ...], dtype[str_]]"
Type of "capitalize(vec_str_or_T)" is "ndarray[tuple[int, ...], StringDType]"
```

The second dtype is wrong: it should've been `dtype[str_] | StringDType` instead.
So effectively, the `list[str]` part of the union was ignored.

---

To solve this case you'll need an extra overload to completely fix this:

```python
@overload
def capitalize(a: U_co) -> NDArray[str_]: ...
@overload
def capitalize(a: S_co) -> NDArray[bytes_]: ...
@overload
def capitalize(a: _SupportsArray[StringDType]) -> AR_T: ...
@overload
def capitalize(a: T_co) -> ndarray[_Shape, dtype[str_] | StringDType]: ...
```

Here, the 3rd overload handles the "only StringDType" case, and the 4th one the "mixed str and StringDType" case

This same fix can also be applied to the other functions whose overloads have incompatible return types, such as `add` and `replace`. 
So the functions that return the same type from each overload, e.g. `split`, don't have this problem.

---

In https://github.com/numpy/numpy/issues/27032#issuecomment-2278958251 I explained this "overlapping overloads" in more detail.

File: numpy/_core/defchararray.pyi [hunk 15]
```diff
@@ -621,49 +680,69 @@ def split(
...
-def splitlines(a: U_co, keepends: None | b_co = ...) -> NDArray[object_]: ...
+def split(
+    a: T_co,
+    sep: None | T_co = ...,
+    maxsplit: None | i_co = ...,
+) -> NDArray[object_]: ...
+
 @overload
-def splitlines(a: S_co, keepends: None | b_co = ...) -> NDArray[object_]: ...
+def splitlines(a: UST_co, keepends: None | b_co = ...) -> NDArray[np.object_]: ...
```
Comment (jorenham) [inline numpy/_core/defchararray.pyi:691]: there's only one overload left now, so the `@overload` can be removed

File: numpy/_core/defchararray.pyi (hunk 14 above)
Comment (jorenham) [reply numpy/_core/defchararray.pyi:542]: BTW, you could also choose to drop the 4th overload in my proposed fix, in case that you think that the additional complexity isn't worth it.

The consequence is that passing `list[str] | ndarray[tuple[int], StringDType]` is effectively disallowed.

Comment (ngoldbaum) [reply numpy/_core/defchararray.pyi:542]: I'm trying to implement this suggestion but got stuck.

For `add` I have:

```python
@overload
def add(x1: U_co, x2: U_co) -> NDArray[np.str_]: ...
@overload
def add(x1: S_co, x2: S_co) -> NDArray[np.bytes_]: ...
@overload
def add(x1: _StringDTypeSupportsArray, x2: _StringDTypeSupportsArray) -> _StringDTypeArray: ...
@overload
def add(x1: T_co, T_co) -> _StringDTypeOrUnicodeArray: ...
```

Where 

```
_StringDTypeArray: TypeAlias = np.ndarray[_Shape, np.dtypes.StringDType]
_T_SA: TypeAlias = _SupportsArray[np.dtypes.StringDType]
_StringDTypeOrUnicodeArray: TypeAlias = np.ndarray[_Shape, np.dtype[np.str_] | np.dtypes.StringDType]
```

That fixes the error you pasted above. Woohoo!

But then for `multiply`, I now have:

```python
@overload
def multiply(a: U_co, i: i_co) -> NDArray[np.str_]: ...
@overload
def multiply(a: S_co, i: i_co) -> NDArray[np.bytes_]: ...
@overload
def multiply(a: _StringDTypeSupportsArray, i: i_co) -> _StringDTypeArray: ...
@overload
def multiply(a: T_co, i: i_co) -> _StringDTypeOrUnicodeArray: ...
```

And now pyright complains:

```
defchararray.pyi:524:5 - error: Overload 2 for "multiply" overlaps overload 4 and returns an incompatible type
```

I guess this is because the second argument is integer? Any idea how to work around this?

**Comment by ngoldbaum** [comment]
I'll finish this off if I can get some help with the test failures and the question I just posted above.

File: numpy/_core/defchararray.pyi (hunk 14 above)
Comment (jorenham) [reply numpy/_core/defchararray.pyi]: I don't think that the seconds argument is relevant here, because it's the same everywhere. Sometimes in cases like these it can help to distribute the dtype over arrays, i.e.

```python
_StringDTypeOrUnicodeArray: TypeAlias = np.ndarray[_Shape, np.dtype[np.str_]] | np.ndarray[_Shape, np.dtypes.StringDType]
```

**Comment by jorenham** [comment]
> I'll finish this off if I can get some help with the test failures and the question I just posted above.

The test failure seems to be a result of an import issue with `StringDType` in `numpy._typing._array_like`:

image

image

So when mypy doesn't know what something is, it just treats is as if its `Any, which explains  why `np.char.equal(AR_U, AR_S)` doesn't fail.

**Comment by ngoldbaum** [comment]
OK, only remaining failure I see is unrelated.

Thanks so much for your help with this @jorenham!

File: numpy/_core/strings.pyi [hunk 16]
```diff
@@ -1,70 +1,103 @@
...
     _ArrayLikeInt_co as i_co,
-    _ArrayLikeBool_co as b_co,
+    _ArrayLikeString_co as T_co,
+    _ArrayLikeAnyString_co as UST_co,
+    _Shape,
+    _SupportsArray,
 )
 
+
+AR_T: TypeAlias = np.ndarray[_Shape, np.dtypes.StringDType]
```
Comment (jorenham) [inline numpy/_core/strings.pyi:16]: It's probably cleaner to rename this using CamelCase convention and add a leading underscore

**Review by jorenham — APPROVED** [review]
Mypy is still reporting some `overload-overlap` errors in `defchararray.pyi`, but I believe those are false positives (i.e. one of those oh-so-familiar mypy bugs), and Pyright seems to be happy with them.

There are some minor opportunities for improvement left, but as far as I'm concerned, those shouldn't be the deal-breakers for merging.
Apart form the review comment I left, I'm referring here to how most overloads could be "merged" through using a `TypeVar` with constraints, but it's probably cleaner to address this in seperate PR now that I think about it 🤔.

Anyway, it's very nice to see that `_core/strings.pyi` is completely error free (which unfortunately is a rare signt in the numpy stubs at the moment).
It's also worth noting that pyright appears to agree with mypy on all relevant type-tests (and that's far from trivial), which is a very good sign.

So thanks for your effort and your patience; typing in Python is tricky, but you seem to have a knack for it 👌🏻.

---

tldr; I'm happy to merge this if you want, so let me me know if you plan on changing anything (e.g. the tags suggest you want to add some release notes?)

**Comment by charris** [comment]
Needs rebase, some of the reveal tests have been removed.

**Comment by ngoldbaum** [comment]
I think this is good to merge now.

**Comment by jorenham** [comment]
> I think this is good to merge now.

I agree 👌🏻

**Comment by jorenham** [comment]
Thanks @ngoldbaum!
